server/src/main/java/org/elasticsearch/health/node/DiskHealthIndicatorService.java

```diff
@@ -244,13 +269,21 @@ private List<HealthIndicatorImpact> getImpacts(
                 )
             );
         }
-        if (redNonDataNonMasterNodes.isEmpty() == false || yellowNonDataNonMasterNodes.isEmpty() == false) {
+        Set<String> unhealthyNonMasterNonDataRoles = unhealthyRoles.stream()
+            .filter(role -> role.canContainData() == false && role.equals(DiscoveryNodeRole.MASTER_ROLE) == false)
+            .map(DiscoveryNodeRole::roleName)
+            .collect(Collectors.toSet());
+        if (unhealthyNonMasterNonDataRoles.isEmpty() == false) {
             impacts.add(
                 new HealthIndicatorImpact(
                     NAME,
                     IMPACT_CLUSTER_FUNCTIONALITY_UNAVAILABLE_ID,
                     2,
-                    "Some cluster functionality might be unavailable.",
+                    String.format(
+                        Locale.ROOT,
+                        "The [%s] functionality might be impaired.",
+                        unhealthyNonMasterNonDataRoles.stream().sorted().collect(Collectors.joining(", "))
```

Comment:
Should we join them above directly in a comma separated String? ( https://github.com/elastic/elasticsearch/pull/90262/files#diff-228b0b1e381859fb6fbbb4655fa4ebd815f6ca0e28a945c3aa66b1f7db2c99c7R275 ) The `unhealthyNonMasterNonDataRoles` set doesn't seem to be used for anything else?

We can test if the string is empty or not to enable this impact (feels a bit wasteful to keep streaming this list)

Reply:
yes I will go through all the cases and see how many of them we can eliminate.